### PR TITLE
Statuses enhancement

### PR DESCRIPTION
**Pull request by Malyuk-A**
# Changes
- Adds a new library that can be imported anywhere.
- This library contains a "types" folder with a file for statuses.
- This file includes many important services, deployment, scheduling-related status
- It should be the single source of truth when it comes to these statuses - thus it is the place to add comments and explanations about when to use what status

# Affected Areas
- These changes only affect the Python code.
- I adjusted every occurrence of (non-HTTP) "status" I could find.

# Disclaimer
These changes are up for debate and are just the first step in bringing order into the chaos.
This MR should not become a huge unmergable beast, thus I want to merge it as soon as possible and use it as a ground for discussion for future improvements that build on top of it.

I have replaced all status strings with Enums however, to not break the running system (especially the HTTP responses and DB interactions, I often times use <status_enum>.value to get the original string value.
This means that there are several places left that should be refactored in follow-up MRs to ideally one day no longer depend on strings and fully support the enums. (E.g. I am using enums for my FLOps Plugin/Extension and because I use plain enums with pydantic I have no issues storing and loading them from/to the DB <- This would require a massive refactoring when applied to Oakestra).

I added many TODOs in the Code to be able to kick-start and follow up with discussions on how to improve things. There are simply too many open questions to resolve all of them at once.

This includes a color-coded Legend in the shared enum statuses file. We should Ideally make sure that we do not have any redundant status as well as have the same set of necessary statuses in Go and Python.

---

Here are a few figures to show what we currently are working with:

Original Mehdi Thesis |  Oakestra Paper 
:-------------------------:|:-------------------------:
  |  

Both are outdated

Go NodeEngine |  Latest 'COMPLETED' state addition
:-------------------------:|:-------------------------:
  |  

Both need polishing and unification.

# Sanity Check
I have always kept a close eye on the system and if it still works. I am able to create apps and services, deploy and delete them, i.e. the normal workflows seem to be working fine. I also keep an eye on the FE.

### CODE REVIEW

File: root_orchestrator/system-manager-python/requirements.txt
Comment (Malyuk-A) [inline root_orchestrator/system-manager-python/requirements.txt:21]: @Malyuk-A change the library dependencies to be pointing to the devel branch before merging

**Comment by Malyuk-A** [comment]
I am looking for feedback on this, especially how I split up/grouped the statuses.

Feel free to suggest other splits and also to respond to open questions/TODOs I have added.
@giobart @TheDarkPyotr @melkodary

File: root_orchestrator/cloud-scheduler/requirements.txt
Comment (melkodary) [inline root_orchestrator/cloud-scheduler/requirements.txt:11]: do we need icecream library?

**Review by melkodary — COMMENTED** [review]
LGTM! Thank you for handling this status mess.

File: root_orchestrator/cloud-scheduler/requirements.txt
Comment (Malyuk-A) [reply root_orchestrator/cloud-scheduler/requirements.txt:11]: Good catch. I use it for debugging. So I will remove it before merging things. TODO @Malyuk-A

File: cluster_orchestrator/cluster-manager/mongodb_client.py
Comment (TheDarkPyotr) [inline cluster_orchestrator/cluster-manager/mongodb_client.py:287]: Just a minor clarification: what do you mean with _previous job_status_? I believe it's coherent if it update the job status from **Cluster Scheduled/Worker Scheduled** to **Failed** (_despite this should not happen considering the 'COMPLETED' state machine diagram_)

Comment (Malyuk-A) [reply cluster_orchestrator/cluster-manager/mongodb_client.py]: The previous job_status is simply the "other status" from a few lines above the "status" 
image

I had another look and managed to resolve the TODO. Now I simply create a new_job_status = FAILED and work with it.

Comment (TheDarkPyotr) [reply cluster_orchestrator/cluster-manager/mongodb_client.py]: You're right, more clear now

**Comment by TheDarkPyotr** [comment]
Thanks for putting this together, looks great! 👍

File: libraries/oakestra_utils_library/oakestra_utils/types/statuses.py
Comment (giobart) [inline libraries/oakestra_utils_library/oakestra_utils/types/statuses.py:55]: In principle:

NODE_SCHEDULED: Container not yet created but node is working on it
ACTIVE:  Container created but task not running yet. Therefore the image was downloaded, container metadata created, but task process inside the container is not running yet. E.g., because we're doing other pre-flight operations such as attaching the network. We can use a more conventional name such as "CREATED", I would personally prefer that. 
RUNNING: Container created and task running

File: libraries/oakestra_utils_library/oakestra_utils/types/statuses.py
Comment (giobart) [inline libraries/oakestra_utils_library/oakestra_utils/types/statuses.py:62]: In theory, but as we know theory and practice don't always match:

FAILED: exit status != 0
DEAD: exist status == 0

File: root_orchestrator/system-manager-python/requirements.txt
Comment (giobart) [reply root_orchestrator/system-manager-python/requirements.txt:21]: If they manually point to develop, how do we make sure that stable branch points to main?

**Review by giobart — CHANGES_REQUESTED** [review]

File: root_orchestrator/system-manager-python/requirements.txt
Comment (Malyuk-A) [reply root_orchestrator/system-manager-python/requirements.txt:21]: That is a great question.

This is currently already used by the Resource Abstractor Client Library. (@melkodary )
I currently see two solutions:
1) Having a dedicated MR for migrating things from devel to stable - in this MR the links need to be adjusted, this is not too pretty.
2) Publish these libraries as pip packages with proper versioning, same redundancy - "headaches" follow. 

I recommend discussing this in a maintainers meeting to come up with a nice convention.

File: libraries/oakestra_utils_library/oakestra_utils/types/statuses.py
Comment (Malyuk-A) [reply libraries/oakestra_utils_library/oakestra_utils/types/statuses.py]: Thank you for the clarification, I have adjusted the file accordingly.
I have renamed the ACTIVE status to CREATED.

File: libraries/oakestra_utils_library/oakestra_utils/types/statuses.py
Comment (Malyuk-A) [reply libraries/oakestra_utils_library/oakestra_utils/types/statuses.py]: Nice clarification. I added explanatory comments.

**Comment by giobart** [comment]
Is it ready to be tested?

**Comment by Malyuk-A** [comment]
> Is it ready to be tested?
@giobart 

I just tested it again locally and it works for me.
So feel free and test it.

Note. Before Merging I need to adjust the library imports to point to develop

**Review by giobart — APPROVED** [review]
image

LGTM!

File: root_orchestrator/system-manager-python/requirements.txt
Comment (melkodary) [reply root_orchestrator/system-manager-python/requirements.txt:21]: I believe the correct approach is the second solution proposed. However for the time being I think there is another solution that could be implemented - create the PR for library and merge it before merging the stuff of the main PR. This way no need to change the links everytime. In other words create two PRs: 
1. one for the new library or modification to the library (assuming changes don't break anything) and merge it.
2. Next create the main PR that uses this library.

**Pull request merged**
